### PR TITLE
Use dora-ts part #1

### PR DESCRIPTION
**Pull request by ixje**
The idea of this PR is to make Dora always make use of [dora-sdk-ts](https://github.com/CityOfZion/dora-sdk-ts) instead of using `dora-sdk-ts` in some parts and other parts manual calling end points using `fetch` and `GENERATE_BASE_URL()`
https://github.com/CityOfZion/dora/blob/778cdc51b4b7ce8158fade6fd25a8125e6d450ef/src/constants.tsx#L97-L108

This should make switching to the Dora v2 backend (once ready) a matter of updating `dora-sdk-ts` version and that's it.

### CODE REVIEW

File: src/actions/contractActions.ts [hunk 1]
```diff
@@ -160,19 +161,14 @@ export function fetchContract(hash: string, populateStates = true) {
...
     if (shouldFetchContract(getState(), hash)) {
       dispatch(requestContract(hash))
-
+      const [network, protocol] = getNetworkAndProtocol()
       try {
-        const response = await fetch(`${GENERATE_BASE_URL()}/contract/${hash}`)
-
-        const json = await response.json()
+        const restAPI = protocol === 'neo3' ? NeoRest : NeoLegacyREST
+        const json: any = await restAPI.contract(hash, network)
```
Comment (ixje) [inline src/actions/contractActions.ts:167]: the `any` type was necessary because otherwise it would complain at this piece a couple lines below
```
json.invocationStats = invocationStats || null
```
because the response doesn't have a `invocationStats` field. 
I guess ideally `requestContractSuccess()` should consume the types from `dora-ts` instead of it's own, but yeah that can be a future enhancement.

File: src/utils/chain.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { treatNetwork } from '../constants'
+import { store } from '../store'
+
+export function getNetworkAndProtocol(): [string, string] {
```
Comment (comountainclimber) [inline src/utils/chain.ts:4]: nitpick: this could be a bit more concise
```javascript
const { network, chain } = store.getState().network
return [treatNetwork(network), chain]
```

File: src/actions/transactionActions.ts [hunk 3]
```diff
@@ -118,25 +118,26 @@ export function fetchTransaction(hash: string, chain: string) {
...
+        )
+        transactionRequestPromises.push(NeoLegacyREST.log(hash, network))
         transactionRequestPromises.push(
-          fetch(`${GENERATE_BASE_URL()}/transaction_abstracts/${hash}`),
+          NeoLegacyREST.transactionAbstracts(hash, network),
         )
+      }
 
       try {
         const responses = await Promise.all(transactionRequestPromises)
```
Comment (comountainclimber) [inline src/actions/transactionActions.ts]: nitpick: we can make this a little verbose with something like this:
```javascript
const [network] = getNetworkAndProtocol()
const responses = await Promise.all(
  chain === "neo3"
    ? [NeoRest.transaction(hash, network), NeoRest.log(hash, network)]
    : [
        NeoLegacyREST.transaction(hash, network),
        NeoLegacyREST.log(hash, network),
        NeoLegacyREST.transactionAbstracts(
          hash,
          network
        ),
      ]
);
```

**Comment by comountainclimber** [comment]
Overall looks good to me thanks for taking on this initiative 💪

File: src/actions/contractActions.ts (hunk 1 above)
Comment (comountainclimber) [reply src/actions/contractActions.ts:167]: Can we just use the optional type here for that field => https://www.typescriptlang.org/docs/handbook/interfaces.html#optional-properties ??

Also I totally agree that types should be exported from the sdk

File: src/actions/transactionActions.ts (hunk 3 above)
Comment (ixje) [reply src/actions/transactionActions.ts]: I couldn't feed this it directly into `Promise.all` due to a massive type error. Squeezing it into a separate variable did work. Still cleaner than before, thanks for the idea 👍

File: src/actions/contractActions.ts (hunk 1 above)
Comment (ixje) [reply src/actions/contractActions.ts:167]: The field is originally not on the response object from the `restAPI.contract()` call (being [this for N3](https://github.com/CityOfZion/dora-sdk-ts/blob/dc462350340fcd60c06c0f96064bbb7007a1e7a9/src/interfaces/api/neo/responses.ts#L74) and [this for N2](https://github.com/CityOfZion/dora-sdk-ts/blob/dc462350340fcd60c06c0f96064bbb7007a1e7a9/src/interfaces/api/neo_legacy/interface.ts#L33)). It is just data from another RPC call glued into the previous response object because it probably needs this at some point. Even the resulting data in the `json` variable that is passed to this 

https://github.com/CityOfZion/dora/blob/778cdc51b4b7ce8158fade6fd25a8125e6d450ef/src/actions/contractActions.ts#L35-L36 

doesn't even match the `Contract` type. The most obvious one is the `invocationStats` property that doesn't exist on it. It seems to somehow work but I fail to understand how this can be type safe 😅 

tldr; i can't make `invocationStats` optional because it didn't exist on the response object to being with.

**Comment by ixje** [comment]
close this in favour of a different approach/new direction
